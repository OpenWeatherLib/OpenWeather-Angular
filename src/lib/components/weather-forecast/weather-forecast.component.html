<span class="ga-display-inline-block">
  <mat-card class="ga-card-forecast-header" *ngIf="forecastWeather">
    <mat-card-header>
      <img mat-card-avatar class="ga-card-header-image" src="{{mostWeatherCondition?.icon}}" alt="Forecast weather pic">
      <mat-card-title>Forecast Weather</mat-card-title>
      <mat-card-subtitle>{{mostWeatherCondition?.description}}</mat-card-subtitle>
    </mat-card-header>
    <img mat-card-image src="{{mostWeatherCondition?.wallpaper}}" alt="Forecast weather pic">
  </mat-card>
</span>

<span class="ga-display-inline-block">
  <div class="ga-forecast-container" *ngIf="forecastWeather">
    <form class="ga-margin-1-rem">
      <mat-form-field class="ga-width-100-per">
        <input matInput placeholder="Search" [(ngModel)]="forecastWeatherSearch" [ngModelOptions]="{standalone: true}"
          (keyup)="searchForecastWeather()">
        <button mat-button matSuffix mat-icon-button aria-label="Clear" (click)="forecastWeatherSearch='';searchForecastWeather();">
          <mat-icon svgIcon="close"></mat-icon>
        </button>
      </mat-form-field>
    </form>

    <mat-list>
      <mat-list-item *ngFor="let forecastPart of forecastWeatherList" class="ga-card-forecast-weather-height-fix">
        <mat-card class="ga-card">
          <mat-card-header>
            <img mat-card-avatar class="ga-card-header-image" src="{{forecastPart?.weatherCondition?.icon}}" alt="Forecast weather pic">
            <mat-card-title>{{forecastPart?.dt_txt}}</mat-card-title>
            <mat-card-subtitle>{{forecastPart?.weatherCondition?.description}}</mat-card-subtitle>
          </mat-card-header>
          <mat-card-content>
            <p>Temperature: {{forecastPart?.main?.temp}}°C
              ({{forecastPart?.main?.temp_min}}°C-{{forecastPart?.main?.temp_max}}°C)</p>
            <p>Pressure: {{forecastPart?.main?.pressure}} mBar</p>
            <p>Humidity: {{forecastPart?.main?.humidity}} %</p>
          </mat-card-content>
        </mat-card>
      </mat-list-item>
    </mat-list>
  </div>
</span>
