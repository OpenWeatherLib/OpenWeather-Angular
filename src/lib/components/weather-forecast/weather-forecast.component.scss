@import '../../../styles/ga-common';

.ga-card-forecast-header {
  @extend .ga-card;
  margin-left: 2rem;
}

.ga-forecast-container {
  height: 90vh;
  overflow: overlay;
  width: 27rem;
}

// Default height for mat-list-item is 48px, but our cards have another height
.ga-card-forecast-weather-height-fix {
  height: 100% !important;
}
