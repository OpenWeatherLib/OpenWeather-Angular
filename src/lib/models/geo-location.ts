import { classJsonKey } from "../decorator";

@classJsonKey("", "")
export class GeoLocation {
  lat: number = 720.0;
  long: number = 720.0;

  isDefault(): boolean {
    return this.lat === 720.0 && this.long === 720.0;
  }
}
