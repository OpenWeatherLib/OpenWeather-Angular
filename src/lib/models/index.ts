export * from "./address-component";
export * from "./city";
export * from "./city2";
export * from "./clouds";
export * from "./coordinates";
export * from "./coordinates2";
export * from "./geometry";
export * from "./main";
export * from "./rain";
export * from "./sys";
export * from "./weather-current";
export * from "./weather-forecast";
export * from "./weather-forecast-part";
export * from "./weather-part";
export * from "./wind";
export * from "./uv-index";
