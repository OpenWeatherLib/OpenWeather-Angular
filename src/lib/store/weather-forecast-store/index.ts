export * from "./weather-forecast.actions";
export * from "./weather-forecast.effects";
export * from "./weather-forecast.reducer";
export * from "./weather-forecast.selectors";
export * from "./weather-forecast.state";
