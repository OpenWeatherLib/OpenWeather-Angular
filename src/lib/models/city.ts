import { GeoLocation } from "./geo-location";
import { classJsonKey } from "../decorator";

@classJsonKey("", "")
export class City {
  id: number = 0;
  name: string = "";
  country: string = "";
  population: number = 0;
  geoLocation: GeoLocation = new GeoLocation();

  isDefault(): boolean {
    return this.id === 0 && this.name === "" && this.country === "" && this.population === 0 && this.geoLocation.isDefault();
  }
}
