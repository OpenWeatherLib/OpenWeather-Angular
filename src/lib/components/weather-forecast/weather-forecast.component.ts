import { Component, OnInit } from "@angular/core";
import { Store } from "@ngrx/store";
import { Observable } from "rxjs";

import WeatherCondition from "@lib/enums/weather-condition.enum";
import { WeatherForecastPart } from "@lib/models";
import { selectIsLoading, selectMostWeatherCondition, selectWeatherForecastList, setFilterRequestAction } from "@lib/store/weather-forecast-store";
import { RootState } from "@lib/store/root-state";

@Component({
  selector: "ga-weather-forecast",
  templateUrl: "./weather-forecast.component.html",
  styleUrls: ["./weather-forecast.component.scss"]
})
export class WeatherForecastComponent implements OnInit {

  isLoading$: Observable<boolean>;

  mostWeatherCondition$: Observable<WeatherCondition>;

  weatherForecastList$: Observable<WeatherForecastPart[]>;

  weatherForecastSearch: string = "";

  readonly filterWeatherForecast = (): void => this.store$.dispatch(setFilterRequestAction({ filter: this.weatherForecastSearch }));

  readonly trackByIndex = (index: number, _: WeatherForecastPart) => index;

  constructor(private readonly store$: Store<RootState>) { }

  ngOnInit() {
    this.isLoading$ = this.store$.select(selectIsLoading);

    this.mostWeatherCondition$ = this.store$.select(selectMostWeatherCondition);

    this.weatherForecastList$ = this.store$.select(selectWeatherForecastList);
  }
}
