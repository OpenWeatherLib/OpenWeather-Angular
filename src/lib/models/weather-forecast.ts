import { classJsonKey } from "../decorator";
import { City } from "./city";
import { WeatherForecastPart } from "./weather-forecast-part";

@classJsonKey(String().empty, String().empty)
export class WeatherForecast {
    city: City = new City();
    list: WeatherForecastPart[] = [];

    isDefault(): boolean {
        return this.city.isDefault() && this.list.length === 0;
    }
}
