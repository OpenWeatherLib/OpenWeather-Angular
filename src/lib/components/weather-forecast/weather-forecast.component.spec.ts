import { NO_ERRORS_SCHEMA } from "@angular/core";
import { TestBed } from "@angular/core/testing";
import { combineReducers, Store, StoreModule } from "@ngrx/store";
import { of } from "rxjs";

import WeatherCondition from "@lib/enums/weather-condition.enum";
import { WeatherForecastPart } from "@lib/models";
import { weatherForecastReducer, setFilterRequestAction } from "@lib/store/weather-forecast-store";
import { RootState } from "@lib/store/root-state";
import { WeatherForecastComponent } from "./weather-forecast.component";

describe("WeatherForecastComponent", () => {
  let component: WeatherForecastComponent;
  let store$: Store<RootState>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [
        WeatherForecastComponent
      ],
      imports: [
        StoreModule.forRoot({ weatherForecast: combineReducers(weatherForecastReducer) })
      ],
      schemas: [NO_ERRORS_SCHEMA]
    }).compileComponents();

    store$ = TestBed.get(Store);
    store$.dispatch = jest.fn();
    store$.select = jest.fn()
      .mockReturnValueOnce(of(false))
      .mockReturnValueOnce(of(WeatherCondition.null))
      .mockReturnValueOnce(of([]));

    const fixture = TestBed.createComponent(WeatherForecastComponent);
    component = fixture.debugElement.componentInstance;
  });

  test("should create the app", () => {
    // Assert
    expect(component).toBeTruthy();
  });

  describe("ngOnInit", () => {
    test("isLoading$", done => {
      // Arrange 
      component.ngOnInit();

      // Act
      component.isLoading$
        .subscribe({
          next: (isLoading: boolean) => {
            // Assert
            expect(isLoading).toBeFalsy();
          },
          complete: () => done()
        });
    });

    test("mostWeatherCondition$", done => {
      // Arrange 
      component.ngOnInit();

      // Act
      component.mostWeatherCondition$
        .subscribe({
          next: (weatherCondition: WeatherCondition) => {
            // Assert
            expect(weatherCondition).toMatchSnapshot({
              id: 0,
              description: "Null",
              wallpaper: "/assets/weather_wallpaper_dummy.png",
              icon: "/assets/weather_dummy.png",
              count: 0
            });
          },
          complete: () => done()
        });
    });

    test("weatherForecastList$", done => {
      // Arrange 
      component.ngOnInit();

      // Act
      component.weatherForecastList$
        .subscribe({
          next: (weatherForecastList: WeatherForecastPart[]) => {
            // Assert
            expect(weatherForecastList).toMatchSnapshot([]);
          },
          complete: () => done()
        });
    });
  });

  describe("filterWeatherForecast", () => {
    test("should dispatch action", () => {
      // Arrange
      component.weatherForecastSearch = "Filter";

      // Act
      component.filterWeatherForecast();

      // Assert
      expect(store$.dispatch).toHaveBeenCalledTimes(1);
      expect(store$.dispatch).toHaveBeenCalledWith(setFilterRequestAction({ filter: "Filter" }));
    });
  });

  describe("trackByIndex", () => {
    test("should return expected value", () => {
      // Arrange & Act
      const actual: number = component.trackByIndex(42, undefined);

      // Assert
      expect(actual).toBe(42);
    });
  });
});
