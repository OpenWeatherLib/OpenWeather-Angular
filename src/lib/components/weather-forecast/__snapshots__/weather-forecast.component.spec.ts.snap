// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherForecastComponent ngOnInit mostWeatherCondition$ 1`] = `
Object {
  "count": 0,
  "description": "Null",
  "icon": "/assets/weather_dummy.png",
  "id": 0,
  "wallpaper": "/assets/weather_wallpaper_dummy.png",
}
`;

exports[`WeatherForecastComponent ngOnInit weatherForecastList$ 1`] = `Object {}`;
